core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java

```diff
@@ -0,0 +1,300 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gateway;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.metadata.SnapshotId;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.*;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDeciders;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.test.ElasticsearchAllocationTestCase;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.equalTo;
+
+/**
+ */
+public class PrimaryShardAllocatorTests extends ElasticsearchAllocationTestCase {
+
+    private final ShardId shardId = new ShardId("test", 0);
+    private final DiscoveryNode node1 = newNode("node1");
+    private final DiscoveryNode node2 = newNode("node2");
+    private final DiscoveryNode node3 = newNode("node3");
+    private TestAllocator testAllocator;
+
+    @Before
+    public void buildTestAllocator() {
+        this.testAllocator = new TestAllocator();
+    }
+
+    /**
+     * Verifies that the canProcess method of primary allocation behaves correctly
+     * and processes only the applicable shard.
+     */
+    @Test
+    public void testNoProcessReplica() {
+        ShardRouting shard = TestShardRouting.newShardRouting("test", 0, null, null, null, false, ShardRoutingState.UNASSIGNED, 0, new UnassignedInfo(UnassignedInfo.Reason.CLUSTER_RECOVERED, null));
+        assertThat(testAllocator.needToFindPrimaryCopy(shard, null), equalTo(false));
+    }
+
+    /**
+     * Tests that when async fetch returns that there is no data, the shard will not be allocated.
+     */
+    @Test
+    public void testNoAsyncFetchData() {
+        RoutingAllocation allocation = setupOneNodeOneShard(yesAllocationDeciders());
+        boolean changed = testAllocator.allocateUnassigned(allocation);
+        assertThat(changed, equalTo(false));
+        assertThat(allocation.routingNodes().ignoredUnassigned().size(), equalTo(1));
+        assertThat(allocation.routingNodes().ignoredUnassigned().get(0).shardId(), equalTo(shardId));
+    }
+
+    /**
+     * Tests when the node returns that no data was found for it (-1), it will be moved to ignore unassigned.
+     */
+    @Test
+    public void testNoAllocationFound() {
+        RoutingAllocation allocation = setupOneNodeOneShard(yesAllocationDeciders());
+        testAllocator.addData(node1, -1);
```

Comment:
can we randomize between a store exception and the current code? (alternatively add another test)